### ISSUE TITLE
Add grammar-constrained decoding via logits processors (Outlines / XGrammar)

### ISSUE THREAD

**Issue by krystophny** [computor-org/vllm-mlx#25]
## Summary

vllm-mlx has no grammar-constrained or guided decoding. All structured output (JSON schema, tool calls) relies on prompt engineering + post-hoc parsing. This is the single largest quality gap vs llama.cpp (GBNF grammars) and vLLM CUDA (xgrammar/outlines).

References upstream waybarrios/vllm-mlx#133.

## What llama.cpp / vLLM CUDA do

At each token generation step, a logits processor zeros out tokens that would violate the grammar/schema. This guarantees structurally valid output at the token level -- impossible to emit malformed JSON.

- **llama.cpp**: GBNF grammars via `json_schema_to_grammar`. Lazy grammar triggers for tool calling.
- **vLLM CUDA**: xgrammar (C pushdown automaton, default) or outlines (FSM-based, legacy). Backend flag: `--guided-decoding-backend`.

## Existing infrastructure in vllm-mlx

The `logits_processors` plumbing already exists but is unused:
- `SimpleEngine` passes through to `mlx_lm.stream_generate(logits_processors=...)`
- `BatchedEngine/Scheduler` has full per-request `logits_processors` wiring in `_step()` (scheduler.py:670-688)
- `mlx_lm.generate_step` signature: `logits_processors: List[Callable[[mx.array, mx.array], mx.array]]`

## Integration options

| Backend | MLX support | Performance | Batch support |
|---------|-------------|-------------|---------------|
| **Outlines** | Yes (`outlines.from_mlxlm()`) | FSM-based, moderate | No batch |
| **XGrammar** | `pip install xgrammar[metal]` but uses PyTorch tensors | C pushdown automaton, fast | Yes |

**Recommended**: Start with Outlines (working MLX integration), optionally add XGrammar later.

## Scope

New module `vllm_mlx/guided_decoding.py`:
- `build_logits_processor(schema, backend) -> Callable` factory
- Backend abstraction (outlines initially, xgrammar optional)
- Wire into SimpleEngine and BatchedEngine via `logits_processors` parameter
- Support `response_format: {type: "json_schema", json_schema: {...}}` -> logits processor
- Support `guided_json`, `guided_regex`, `guided_grammar` in `extra_body` (vLLM API compat)
- CLI flag: `--guided-decoding-backend outlines|xgrammar`
- New optional dependency: `outlines`

## Architecture

```
API Layer (server.py)
  -> Guided Decoding Layer (NEW: guided_decoding.py)
    -> Engine Layer (simple.py / scheduler.py)
      -> mlx_lm generate_step (logits_processors hook)
```

## Files
- NEW: `vllm_mlx/guided_decoding.py`
- `vllm_mlx/server.py`
- `vllm_mlx/engine/simple.py`
- `vllm_mlx/engine/batched.py`
- `pyproject.toml` (optional dep)

## Priority
P1 -- high effort but foundational. Issues #26 and #27 depend on this.
